### PR TITLE
plugins/copilot-chat: init

### LINKED ISSUE #992: [PLUGIN REQUEST] CopilotChat.nvim
| Field    | Description |
|----------|-------------|
| Plugin   | `CopilotChat.nvim`  | 
| Homepage | `https://github.com/jellydn/CopilotChat.nvim`     | 
| Nixpkgs  | `false` | 

## Extra Information

Newer features are only present in the separate ``canary`` branch.

Requires four python packages from ``requirements.txt``, all of which are present in nixpkgs.

```
python-dotenv
requests
pynvim==0.5.0
prompt-toolkit
```

Nixpkgs stable has 0.4.3 of pynvim, unstable is at 0.5.0.

This is a fork by jellydn based off of [this](https://github.com/gptlang/CopilotChat.nvim), but the original author also uses this fork now as it is more well maintained.

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for [CopilotChat.nvim](https://github.com/CopilotC-Nvim/CopilotChat.nvim).

Fixes #992

### CODE REVIEW

File: plugins/ai/copilot-chat.nix [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            prompt = "Write commit message for the change with commitizen convention. Make sure the title has maximum 50 characters and message is wrapped at 72 characters. Wrap the whole message in code block with language gitcommit.";
+            selection = ''
+              function(source)
+                return select.gitdiff(source, true)
+              end
+            '';
+          };
+        };
+      in
+      helpers.defaultNullOpts.mkAttrsOf (with types; either str promptType) default ''
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:173]: Should this be `attrsOf (maybeRaw (either str promptType))`?

File: plugins/ai/copilot-chat.nix [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+          end
+        ''
+        ''
+          Default selection (visual or line).
+          `fun(source: CopilotChat.config.source):CopilotChat.config.selection`
+        '';
+
+    prompts =
+      let
+        promptType = types.submodule {
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:117]: Should this have a `freeformType`?

File: plugins/ai/copilot-chat.nix [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            "horizontal"
+            "float"
+            "replace"
+          ]
+          ''
+            Layout for the window.
+          '';
+
+      width =
+        helpers.defaultNullOpts.mkNullable (with types; either (number.between 0.0 1.0) ints.positive) 0.5
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:191]: Should this (and other similar options) accept a raw type too?

Comment (MattSturgeon) [reply plugins/ai/copilot-chat.nix:191]: Maybe we should be reaching for `mkNullableWithRaw` for most custom-type options? 🤔

File: plugins/ai/copilot-chat.nix [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        helpers.defaultNullOpts.mkEnumFirstDefault
+          [
+            "editor"
+            "win"
+            "cursor"
+            "mouse"
+          ]
+          ''
+            Relative position.
+            (Only for floating windows.)
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:212]: Single line breaks don't render. If you wanted an actual line break you'd need to end the line with `\`, or for a full paragraph break you need a blank line.

I think this applies to a few option descriptions.

If they're rendering as you intended then ignore this.

File: plugins/ai/copilot-chat.nix [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        [
+          "buffers"
+          "buffer"
+        ]
+        null
+        ''
+          Default context to use, `"buffers"`, `"buffer"` or `null` (can be specified manually in prompt via @).
+        '';
+
+    history_path = helpers.defaultNullOpts.mkStr (helpers.mkRaw "vim.fn.stdpath('data') .. '/copilotchat_history'") ''
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:92]: Passing a raw type here is tripping `mkStr`'s assertion. It's the assertion that's wrong though.

It should be `assert default == null || isSteing default || isRaw default`. But realistically, we can remove the assertion now that we know `mkDesc` & `toPretty` are working correctly.

File: plugins/ai/copilot-chat.nix [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    allow_insecure = helpers.defaultNullOpts.mkBool false ''
+      Allow insecure server connections.
+    '';
+
+    system_prompt = helpers.defaultNullOpts.mkLua "require('CopilotChat.prompts').COPILOT_INSTRUCTIONS" ''
+      System prompt to use.
+    '';
+
+    model = helpers.defaultNullOpts.mkStr "gpt-4" ''
+      GPT model to use, 'gpt-3.5-turbo' or 'gpt-4'.
```
[the remark this answers is not in the document]
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix:34]: Yes, I am aware of that, and I made this choice on purpose. I suspect the model offering to evolve and I don't feel like constraining this option has a lot of value.

File: plugins/ai/copilot-chat.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix:173]: By default, `helpers.defaultNullOpts.mkAttrsOf` set the type to be `attrsOf (maybeRaw <TYPE>)`:
https://github.com/nix-community/nixvim/blob/d15fade62b743839a20d927d3506d503858f49f0/lib/options.nix#L122

Is that what you were talking about ?

File: plugins/ai/copilot-chat.nix (hunk 3 above)
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix:191]: Yes, good idea. Done.

File: plugins/ai/copilot-chat.nix (hunk 2 above)
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix:117]: Doesn't hurt. Done.

File: plugins/ai/copilot-chat.nix [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+  maintainers = [ maintainers.GaetanLepage ];
+
+  settingsOptions = {
+    debug = helpers.defaultNullOpts.mkBool false ''
+      Enable debug logging.
+    '';
+
+    proxy = helpers.defaultNullOpts.mkStr null ''
+      Custom proxy to use: `[protocol://]host[:port]`.
```
[the remark this answers is not in the document]
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix]: Yes, this is the format the proxy string should follow. I enhanced the comment a bit.

File: plugins/ai/copilot-chat.nix [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+with lib;
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "copilot-chat";
+  originalName = "CopilotChat.nvim";
```
Comment (MattSturgeon) [inline plugins/ai/copilot-chat.nix:11]:
```
 > ERROR: Error detected while processing /nix/store/iw1f7231sl0b0r7z5yxkjyy9sw70z00x-init.lua:
> E5113: Error while calling lua chunk: /nix/store/iw1f7231sl0b0r7z5yxkjyy9sw70z00x-init.lua:11: module 'copilot-chat' not found:
> no field package.preload['copilot-chat']
> no file '/nix/store/nmcclggvanaxcdb84hk1mdqimcfc4gbw-luajit-2.1.1693350652-env/share/lua/5.1/copilot-chat.lua'
> no file '/nix/store/nmcclggvanaxcdb84hk1mdqimcfc4gbw-luajit-2.1.1693350652-env/share/lua/5.1/copilot-chat/init.lua'
> no file '/nix/store/nmcclggvanaxcdb84hk1mdqimcfc4gbw-luajit-2.1.1693350652-env/lib/lua/5.1/copilot-chat.so'
> stack traceback:
> [C]: in function 'require'
> /nix/store/iw1f7231sl0b0r7z5yxkjyy9sw70z00x-init.lua:11: in main chunk
```

```suggestion
  originalName = "CopilotChat.nvim";
  luaName = "CopilotChat";
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
Otherwise looks good!

File: plugins/ai/copilot-chat.nix (hunk 8 above)
Comment (GaetanLepage) [reply plugins/ai/copilot-chat.nix:11]: Good catch

**Review by MattSturgeon — APPROVED** [review]
Looks good!

**Pull request merged**